File: morning_exercises/day26_2.cpp
#include <utility>
#include <vector>
#include <deque>
#include <iostream>

std::vector<std::vector<int>> createGraph(int num, std::vector<std::pair<int, int>>& prereqs) {
    std::vector<std::vector<int>> graph(num);

    for (const auto& [a,b] : prereqs) graph[b].push_back(a);
    return graph;
}

bool dfsCheck(std::vector<std::vector<int>>& graph, int& node, std::vector<int>& state, std::deque<int>& result) {
    if (state[node]==2) return true;
    if (state[node]==1) return false;

    state[node]=1;
    for (int neighbour : graph[node]) {
        if(!dfsCheck(graph, neighbour, state, result)) return false;
    }
    state[node]=2;
    result.push_front(node);
    return true;
}

std::deque<int> isDoable(std::vector<std::vector<int>>& graph) {
    std::vector<int> state(graph.size());

    std::deque<int> result = {};

    for (int i=0;i<graph.size();i++) {
        if (!dfsCheck(graph, i, state, result)) return {};
    }
    return result;
}

int main() {
    std::vector<std::pair<int,int>> prerequisites = {{1,2}, {2,3}};
    auto graph = createGraph(4, prerequisites);

    // std::cout << std::boolalpha;
    auto result = isDoable(graph);
    for (int c : result) {
        std::cout << c << "\n";
    }
}

File: morning_exercises/day26_2.py
from collections import deque

def createGraph(number: int, prerequisites: list)->list:
    graph = [[] for _ in range(0, number)]
    for a,b in prerequisites: graph[b].append(a)
    return graph

def dfsCheck(graph: list, node: int, state: list, result: deque)->bool:
    if state[node]==2: return True
    if state[node]==1: return False

    state[node]=1
    for neighbour in graph[node]:
        if not dfsCheck(graph, neighbour, state, result): return False
    state[node]=2
    result.appendleft(node)
    return True

def isDoable(graph: list)->deque:
    state = [0 for _ in range(len(graph))]
    result = deque()
    for i in range(len(graph)):
        if not dfsCheck(graph, i, state, result): return deque()
    return result

courseschedule = createGraph(2, [[1,0], [0,1]])
print(isDoable(courseschedule))